Replace non-standard CLZ builtins with c++20's bit_width

theuni opened the pull request:
Split out of #28674

Note that we can't yet drop our configure checks because we pass the results on to minisketch. I've opened a PR for that upstream here: https://github.com/sipa/minisketch/pull/80

fanquake [suggested](https://github.com/bitcoin/bitcoin/pull/28674#discussion_r1422215535) that we simply replace our `CountBits` call-sites with uses of `std::bit_width` directly and just drop the tests and fuzzers. I agree with that, but I wanted to allow our tests/fuzzers to run with this change first to help convince reviewers that `std::bit_width` is a drop-in replacement for`CountBits`.

I can either add a commit on top of this PR to do the switch after the c-i has run or do it as a follow-up PR, I have no preference.

I was curious to see what would happen under the hood with this change. Fortunately libc++ (as an example) does exactly what one would expect, using the built-ins: https://github.com/llvm/llvm-project/blob/main/libcxx/include/__bit/countl.h#L56

theuni committed:
crypto: replace non-standard CLZ builtins with c++20's bit_width

theuni committed:
crypto: add BitWidth bench

theStack reviewed (approved):
Code-review ACK 273e4dc8da41f113b21a7ece54988b7bddf3612e

Unit tests succeeded locally :heavy_check_mark: 

Seems reasonable to keep `CountBits` in a first step with its tests and fuzzers for reassurance and then replace all of its call-sites. Happy to re-review this PR or a follow-up, depending on where you decide to put the next commit.

hebasto commented:
Concept ACK.

hebasto commented:
> ... but I wanted to allow our tests/fuzzers to run with this change first to help convince reviewers that `std::bit_width` is a drop-in replacement for`CountBits`.

I'm convinced :)

fanquake commented:
Concept ACK. We can also check the benchmarks in https://corecheck.dev/bitcoin/bitcoin/pulls/29057 when they become available.

> I can either add a commit on top of this PR to do the switch after the c-i has run or do it as a follow-up PR, I have no preference.

I think the changes here should be straightforward enough, that you could push another commit doing the cleanup.

theuni commented:
Thanks @theStack and @hebasto for the quick review.

@fanquake sounds good, will push a commit to do the switch once we see some benchmarks.

theuni committed:
crypto: replace CountBits with std::bit_width

bit_width is a drop-in replacement with an exact meaning in c++, so there is
no need to continue testing/fuzzing/benchmarking.

aureleoules commented:
> I don't think the benchmarks will come. I guess the machine has been preempted by AWS? @aureleoules

I've re-ran the job. I actually found a way to re-run preempted jobs automatically!
A lot of benchmark jobs failed yesterday because of #29061 including this pull because I rebase all pulls on master before running jobs and I don't properly handle failures 😬

aureleoules commented:
So the benchmark results are available now but they are pessimistic on a few benchs (WalletCreateTxUseOnlyPresetInputs, WalletAvailableCoins and PrevectorClearNontrivial).

I ran the benchmarks locally without cachegrind and there does not seem to be any variation between master and the pull.
But when I run them with cachegrind I notice the same variation as seen on corecheck.

This may be due to a limitation of cachegrind because it cannot emulate L2 cache and so results may be pessimist? I'm not an expert in this field though so I'll let you interpret the results😅

theuni commented:
I can reproduce a slight slowdown here. I added a bench that demonstrates the difference, then the original commit, then the removal.

Both libc++ and libstdc++ implement this in terms of the same built-ins we were using before, so I find this surprising. I'd hate to find that the c++ism's have a cost.

Going to convert this to a draft while I look into it.

theuni commented:
@maflcko I'm thinking we should just take the c++20 code and assume it will be more performant going forward. That makes sense to me anyway.

I'm going to go ahead and close this and include it in another PR which gathers the serialization c++20 changes. We can do another round of benches there.
